src/hotspot/share/opto/mempointer.hpp

```diff
@@ -461,98 +536,276 @@ class MemPointerSummand : public StackObj {
+
+#ifndef PRODUCT
+    void print_on(outputStream* st) const {
+      switch (_kind) {
+      case Object:
+          st->print("object  ");
+          st->print("%d %s", _base->_idx, _base->Name());
+          break;
+      case Native:
+          st->print("native  ");
+          st->print("%d %s", _base->_idx, _base->Name());
+          break;
+      default:
+          st->print("unknown ");
+      };
+    }
+#endif
+
+  private:
+    static Node* find_base(Node* object_base, const GrowableArray<MemPointerSummand>& summands);
+  };
 
 private:
+  NOT_PRODUCT( const TraceMemPointer& _trace; )
+  MemPointerSummand _summands[SUMMANDS_SIZE];
+  const NoOverflowInt _con;
+  const Base _base;
+  const jint _size;
+
   // Default / trivial: pointer = 0 + 1 * pointer
-  MemPointerDecomposedForm(Node* pointer) : _pointer(pointer), _con(NoOverflowInt(0)) {
+  MemPointer(Node* pointer,
+             const jint size
+             NOT_PRODUCT(COMMA const TraceMemPointer& trace)) :
+    NOT_PRODUCT(_trace(trace) COMMA)
+    _con(NoOverflowInt(0)),
+    _base(Base()),
+    _size(size)
+  {
     assert(pointer != nullptr, "pointer must be non-null");
     _summands[0] = MemPointerSummand(pointer, NoOverflowInt(1));
+    assert(1 <= _size && _size <= 2048 && is_power_of_2(_size), "valid size");
   }
 
-  MemPointerDecomposedForm(Node* pointer, const GrowableArray<MemPointerSummand>& summands, const NoOverflowInt& con)
-    : _pointer(pointer), _con(con) {
+  // pointer = SUM(SUMMANDS) + con
+  MemPointer(Node* pointer,
+             const GrowableArray<MemPointerSummand>& summands,
+             const NoOverflowInt& con,
+             const jint size
+             NOT_PRODUCT(COMMA const TraceMemPointer& trace)) :
+    NOT_PRODUCT(_trace(trace) COMMA)
+    _con(con),
+    _base(Base::make(pointer, summands)),
+    _size(size)
+  {
     assert(!_con.is_NaN(), "non-NaN constant");
     assert(summands.length() <= SUMMANDS_SIZE, "summands must fit");
+#ifdef ASSERT
     for (int i = 0; i < summands.length(); i++) {
-      MemPointerSummand s = summands.at(i);
+      const MemPointerSummand& s = summands.at(i);
       assert(s.variable() != nullptr, "variable cannot be null");
       assert(!s.scale().is_NaN(), "non-NaN scale");
-      _summands[i] = s;
+    }
+#endif
+
+    // Put the base in in the 0th summand.
+    Node* base = _base.object_or_native_or_null();
+    int pos = 0;
+    if (base != nullptr) {
+      MemPointerSummand b(base, NoOverflowInt(1));
+      _summands[0] = b;
+      pos++;
+    }
+    // Put all other summands afterward.
+    for (int i = 0; i < summands.length(); i++) {
+      const MemPointerSummand& s = summands.at(i);
+      if (s.variable() == base && s.scale().is_one()) { continue; }
+      _summands[pos++] = summands.at(i);
+    }
+    assert(pos == summands.length(), "copied all summands");
+
+    assert(1 <= _size && _size <= 2048 && is_power_of_2(_size), "valid size");
```

Comment:
Where 2048 comes from? Do you have a runtime check somewhere too?

Reply:
It is just a sanity check: currently no platform has vectors larger than 2048. Replaced the assert with:
`assert(1 <= _size && _size <= 2048 && is_power_of_2(_size), "sanity: no vector is expected to be larger");`

Reply:
Okay.